rslib/src/search/writer.rs

```diff
@@ -0,0 +1,272 @@
+// Copyright: Ankitects Pty Ltd and contributors
+// License: GNU AGPL, version 3 or later; http://www.gnu.org/licenses/agpl.html
+
+use crate::{
+    backend_proto::concatenate_searches_in::Separator,
+    decks::DeckID as DeckIDType,
+    err::{AnkiError, Result},
+    notetype::NoteTypeID as NoteTypeIDType,
+    search::parser::{parse, Node, PropertyKind, SearchNode, StateKind, TemplateKind},
+};
+use itertools::Itertools;
+use std::mem;
+
+/// Take an Anki-style search string and convert it into an equivalent
+/// search string with normalized syntax.
+pub fn normalize_search(input: &str) -> Result<String> {
+    Ok(write_nodes(&parse(input)?))
+}
+
+/// Take an Anki-style search string and return the negated counterpart.
+/// Empty searches (whole collection) remain unchanged.
+pub fn negate_search(input: &str) -> Result<String> {
+    let mut nodes = parse(input)?;
+    use Node::*;
+    Ok(if nodes.len() == 1 {
+        let node = nodes.remove(0);
+        match node {
+            Not(n) => write_node(&n),
+            Search(SearchNode::WholeCollection) => "".to_string(),
+            Group(_) | Search(_) => write_node(&Not(Box::new(node))),
+            _ => unreachable!(),
+        }
+    } else {
+        write_node(&Not(Box::new(Group(nodes))))
+    })
+}
+
+/// Take arbitrary Anki-style search strings and return their concatenation where they
+/// are separated by the provided boolean operator.
+/// Empty searches (whole collection) are left out.
+pub fn concatenate_searches(sep: i32, searches: &[String]) -> Result<String> {
+    let bool_node = vec![match Separator::from_i32(sep) {
```

Comment:
It would be cleaner to define a native enum for sep, and convert from the Protobuf enum to the native one in the backend methods. That way if other parts of the Rust code want to call this routine, they don't need to  bother with Protobuf. You'll find a bunch of 'impl From' lines in backend/mod.rs so we can take a protobuf value and '.into()' it to get a native equivalent. I started using the from/into stuff part-way through - an older example of enum conversion is sort_kind_from_pb()